src/ra_log.erl

```diff
@@ -467,37 +470,59 @@ fetch_term(Idx, #state{cache = Cache, uid = UId} = State0) ->
             end
     end.
 
--spec install_snapshot(Snapshot :: ra_log:snapshot(),
+-spec install_snapshot(Meta :: ra_snapshot:meta(),
+                       Data :: term(),
                        State :: ra_log()) ->
-    ra_log().
-install_snapshot({Idx, Term, _, _} = Snapshot,
-                 #state{directory = Dir} = State) ->
+    {ra_log(), term()}.
+install_snapshot({Idx, Term, _} = Meta, Data,
+                 #state{directory = Dir,
+                        snapshot_module = SnapModule} = State) ->
     % syncronous call when follower receives a snapshot
-    {ok, File, Old} = ra_log_snapshot_writer:write_snapshot_call(Dir, Snapshot),
-    handle_event({snapshot_written, {Idx, Term}, File, Old},
-                 State#state{last_index = Idx,
-                             last_written_index_term = {Idx, Term}}).
+    {ok, File, Old} = ra_log_snapshot_writer:save_snapshot_call(Dir, Meta, Data, SnapModule),
+    {ok, MacState} = SnapModule:install(Data, File),
+    State1 = handle_event({snapshot_written, {Idx, Term}, File, Old},
+                          State#state{last_index = Idx,
+                                      last_written_index_term = {Idx, Term}}),
+    {State1, MacState}.
 
 -spec read_snapshot(State :: ra_log()) ->
-    maybe(ra_log:snapshot()).
+    maybe({ok, ra_snapshot:meta(), term()}).
 read_snapshot(#state{snapshot_state = undefined}) ->
     undefined;
-read_snapshot(#state{snapshot_state = {_, _, File}}) ->
-    case ra_log_snapshot:read(File) of
-        {ok, Snapshot} ->
-            Snapshot;
+read_snapshot(#state{snapshot_state = {_, _, File},
+                     snapshot_module = SnapModule}) ->
+    case SnapModule:read(File) of
+        {ok, Meta, Data} ->
+            {ok, Meta, Data};
         {error, enoent} ->
             undefined
     end.
 
+-spec recover_snapshot(State :: ra_log()) ->
+    maybe({ok, ra_snapshot:meta(), term()}).
+recover_snapshot(#state{snapshot_state = undefined}) ->
+    undefined;
+recover_snapshot(#state{snapshot_state = {_, _, File},
+                        snapshot_module = SnapModule}) ->
+    case SnapModule:recover(File) of
+        {ok, Meta, Data} ->
+            {ok, Meta, Data};
+        {error, enoent} ->
```

Comment:
should we handle this? when would we have a snapshot state but no file on disk?

Reply:
When the node is starting, recover expects undefined. This behaviour was there before, I just moved it to this function.